src/capmap.cc

```diff
@@ -232,4 +238,38 @@ bool LoadMap::try_combine(void* __capability cap) {
   return false;
 }
 
+bool BasicMap::try_combine(void* __capability cap) {
+  if (!cheri_tag_get(cap) || cheri_is_sealed(cap)) return false;
+  if ((cheri_perms_get(cap) & perms_) != perms_) return false;  // match ALL perms
+
+  ranges_.combine(Range::from_cap(cap));
+  return true;
+}
+
+bool BranchMap::try_combine(void* __capability cap) {
+  if (!cheri_tag_get(cap) || cheri_type_get(cap) > CHERI_OTYPE_SENTRY) return false;
```

Comment:
The condition `cheri_type_get(cap) > CHERI_OTYPE_SENTRY` is intended to return false if the cap is *not* a possible direct branch target (either an unsealed cap or a sentry), since sentries are otype 1 and unsealed caps are otype 0. I see your point about portability -- would you prefer the (for Morello equivalent) `(cheri_is_sealed(cap) && !cheri_is_sentry(cap))`?

I took the decision in this map to exclude LB, LPB, and sealed pairs. While they are potential branch targets, the inclusion of a data cap to install makes them logically distinct, so I would expect a separate map would be appropriate. In the case of LB and LPB, the branch address can't be directly inspected anyway. Would it be better if I named this class `DirectBranchMap` or `SimpleBranchMap` to indicate that the c29-tied branches are excluded?

Reply:
Yep, we should consider `CHERI_OTYPE_SENTRY` to be opaque, unless some CHERI C reference states otherwise. I can't remember the semantics of `cheri_is_sentry()` regarding the other reserved otypes, but if it's only true for `OTYPE_SENTRY`/`RB`, then yes, that's good :-)

> I took the decision in this map to exclude LB, LPB, and sealed pairs.

Yep, they need special handling, and might not even make sense in this framework. They're typically compartment transitions, so perhaps we could mark them (with a separate `Map`) as external compartment entries. It's probably not a concern for `BranchMap`.

> Would it be better if I named this class

`BranchMap` seems fine to me.

Reply:
I think so!